File: emotiv-sdk/examples_basic/C++/MentalCommandWithCloudProfile/MentalCommandWithCloudProfile.cpp
/****************************************************************************
**
** Copyright 2016 by Emotiv. All rights reserved
** Example - Mental command with  cloud profile
** This example demonstrates how to detect mental command detection with 
** a profile.
** This example work on single connection
****************************************************************************/

#include <stdio.h>
#include <iostream>
#include <sstream>
#include <thread>

#include "IEmoStateDLL.h"
#include "Iedk.h"
#include "IedkErrorCode.h"
#include "EmotivCloudClient.h"
#include "MentalCommandDetection.h"

#ifdef _WIN32
#include <conio.h>
#include <windows.h>
#pragma comment(lib, "Ws2_32.lib")
#endif
#if __linux__ || __APPLE__
    #include <unistd.h>
    #include <termios.h>
    int _kbhit(void);
    int _getch( void );
#endif


const std::string userName = "YourEmotivID";
const std::string password = "YourPassword";
const std::string profileName = "Profile_1";

unsigned int userID = 0;
int userCloudID = -1;
bool enableLoger = false;

int version	= -1; // Lastest version
bool running = true;
typedef unsigned long ulong;
void help();

void SavingLoadingFunction(int userCloudID, bool save = true)
{
	int getNumberProfile = EC_GetAllProfileName(userCloudID);
	
	if (save) {    // Save profile to cloud
        int profileID = -1;
        EC_GetProfileId(userCloudID, profileName.c_str(), &profileID);

		if (profileID >= 0) {
			std::cout << "Profile with " << profileName << " is existed." << std::endl;
            if (EC_UpdateUserProfile(userCloudID, userID, profileID) == EDK_OK)
				std::cout << "Updating finished" << std::endl;
			else 
				std::cout << "Updating failed" << std::endl;
		}
        else if (EC_SaveUserProfile(userCloudID, userID, profileName.c_str(), TRAINING) == EDK_OK)
		{
			std::cout << "Saving finished" << std::endl;
		}
		else std::cout << "Saving failed" << std::endl;
		return;
	} else { // Load profile from cloud
		if (getNumberProfile > 0){
            if (EC_LoadUserProfile(userCloudID, userID, EC_ProfileIDAtIndex(userCloudID, 0)) == EDK_OK)
				std::cout << "Loading finished" << std::endl;
			else
				std::cout << "Loading failed" << std::endl;

		}
		return;
	}
}

void handleMentalCommandEvent(unsigned int userID, EmoEngineEventHandle cognitivEvent)
{
    IEE_MentalCommandEvent_t eventType = IEE_MentalCommandEventGetType(cognitivEvent);

	switch (eventType) {

    case IEE_MentalCommandTrainingStarted:
    {
        std::cout<< std::endl << "MentalCommand training for user " << userID
           << " STARTED!" << std::endl;
        break;
    }

    case IEE_MentalCommandTrainingSucceeded:
    {
        std::cout<< std::endl << "MentalCommand training for user " << userID
           << " SUCCEEDED!" << std::endl;		
		std::cout << "Accept Training!!!" << std::endl;
		IEE_MentalCommandSetTrainingControl(userID, MC_ACCEPT);
		// Reject Training:
		// IEE_MentalCommandSetTrainingControl(userID, MC_REJECT);
        break;
    }

    case IEE_MentalCommandTrainingFailed:
    {
        std::cout<< std::endl << "MentalCommand training for user " << userID
           << " FAILED!" << std::endl;
        break;
    }

    case IEE_MentalCommandTrainingCompleted:
    {
        std::cout<< std::endl << "MentalCommand training for user " << userID
           << " COMPLETED!" << std::endl;
        break;
    }

    case IEE_MentalCommandTrainingDataErased:
    {
        std::cout<< std::endl << "MentalCommand training data for user " << userID
           << " ERASED!" << std::endl;
        break;
    }

    case IEE_MentalCommandTrainingRejected:
    {
        std::cout<< std::endl << "MentalCommand training for user " << userID
           << " REJECTED!" << std::endl;
        break;
    }

    case IEE_MentalCommandTrainingReset:
    {
        std::cout<< std::endl << "MentalCommand training for user " << userID
           << " RESET!" << std::endl;
        break;
    }

    case IEE_MentalCommandAutoSamplingNeutralCompleted:
    {
        std::cout<< std::endl << "MentalCommand auto sampling neutral for user "
           << userID << " COMPLETED!" << std::endl;
        break;
    }

    case IEE_MentalCommandSignatureUpdated:
    {
        std::cout << std::endl << "MentalCommand signature for user " << userID
           << " UPDATED!" << std::endl;
        break;
    }

    case IEE_MentalCommandNoEvent:
        break;

    default:
        break;
	}
}

static bool handleUserInput()
{
	static std::string inputBuffer;
	char key = _getch();

	switch (key)
	{
	case '1':
	{
        int getNumberProfile = EC_GetAllProfileName(userCloudID);

        int profileOK = false;

        int profileID = -1;
        EC_GetProfileId(userCloudID, profileName.c_str(), &profileID);

        if (profileID >= 0) {
            if (EC_LoadUserProfile(userCloudID, userID, profileID) == EDK_OK) {
                profileOK = true;
            }
        }
        if (profileOK == false) {
            if (EC_SaveUserProfile(userCloudID, userID, profileName.c_str(), TRAINING) == EDK_OK)
            {
                std::cout << "Saving finished" << std::endl;
            }
            else std::cout << "Saving failed" << std::endl;
        }

		ulong action1 = (ulong)IEE_MentalCommandAction_t::MC_LEFT;
		ulong action2 = (ulong)IEE_MentalCommandAction_t::MC_RIGHT;
		ulong listAction = action1 | action2;
		int errorCode = EDK_OK;
		errorCode = IEE_MentalCommandSetActiveActions(userID, listAction);
		if (errorCode == EDK_OK)
		{
			std::cout << "Setting MentalCommand active actions (MC_LEFT | MC_RIGHT) for user " << userID << std::endl;			
		}
		else std::cout << "Setting MentalCommand active actions error: " << errorCode;
	}
	break;
	case '2':
		std::cout << std::endl << "Starting training NEUTRAL !" << std::endl;
		IEE_MentalCommandSetTrainingAction(userID, MC_NEUTRAL);
		IEE_MentalCommandSetTrainingControl(userID, MC_START);
		break;
	case '3':
		std::cout << std::endl << "Starting training RIGHT !" << std::endl;
		int errorCode;
		errorCode = IEE_MentalCommandSetTrainingAction(userID, MC_RIGHT);
		errorCode = IEE_MentalCommandSetTrainingControl(userID, MC_START);
		break;
	case '4':
		std::cout << std::endl << "Starting training LEFT !" << std::endl;
		IEE_MentalCommandSetTrainingAction(userID, MC_LEFT);
		IEE_MentalCommandSetTrainingControl(userID, MC_START);
		break;
	case '5':
	{
		int level = 0;
		if (IEE_MentalCommandSetActivationLevel(userID, 2) == EDK_OK) {
			IEE_MentalCommandGetActivationLevel(userID, &level);
			std::cout << "Set MentalCommand Activation level to " << level << std::endl;
		}
		else {
			std::cout << "Set MentalCommand Activation level failured " << level << std::endl;
		}
	}
	break;
	case '6':
	{
		int level = 0;
		IEE_MentalCommandGetActivationLevel(userID, &level);
		std::cout << "Current MentalCommand Activation level: " << level << std::endl;
	}
		break;
	case '7':
	{
		float skill = 0;
		IEE_MentalCommandGetOverallSkillRating(userID, &skill);
		std::cout << "Current overall skill rating: " << skill << std::endl;
	}
	break;
	case '8':
		std::cout << "Saving Profile ... " <<  std::endl;
		SavingLoadingFunction(userCloudID, true);
		break;
	case '9':
		std::cout << "Loading Profile ... " << std::endl;
		SavingLoadingFunction(userCloudID, false);
		break;
	case '0':
		help();
		break;
	case '`':
		return false;
		break;
	default:
		break;
	}
	return true;
}



int main(int argc, char** argv)
{
	int option	= 0;
	std::string input;
	bool ready  = false;

	EmoEngineEventHandle eEvent	= IEE_EmoEngineEventCreate();
	EmoStateHandle eState		= IEE_EmoStateCreate();
	
	int state			        = 0;		
		
	if (IEE_EngineConnect() != EDK_OK) {
        std::cout << "Emotiv Driver start up failed.";
        return -1;
	}

	std::cout << "===================================================================" << std::endl;
    std::cout << "Example to show how to detect mental command detection with a profile. " << std::endl;
    std::cout << "===================================================================" << std::endl;
	std::cout << "Connecting to cloud ..." << std::endl;

    if(EC_Connect() != EDK_OK)
	{
		std::cout << "Cannot connect to Emotiv Cloud";
        return -1;
	}

    if(EC_Login(userName.c_str(), password.c_str()) != EDK_OK)
	{			
		std::cout << "Your login attempt has failed. The username or password may be incorrect" << std::endl;
#ifdef _WIN32
        _getch();
#endif
        return -1;
	}

	std::cout<<"Logged in as " << userName << std::endl;

    if (EC_GetUserDetail(&userCloudID) != EDK_OK) {
        std::cout << "Error: Cannot get detail info, exit " << userName << std::endl;
        return -1;
    }

	help();

	while (true) {
		
		if (_kbhit()) {
			if (!handleUserInput()) {
				break;
			}
		}

		int state = IEE_EngineGetNextEvent(eEvent);

		if (state == EDK_OK) {

			IEE_Event_t eventType = IEE_EmoEngineEventGetType(eEvent);
			IEE_EmoEngineEventGetUserId(eEvent, &userID);

			switch (eventType) 
			{
			    case IEE_UserAdded: 
					std::cout << std::endl << "New user " << userID << " added" << std::endl;
					break;
				case IEE_UserRemoved:
					std::cout << std::endl << "User " << userID << " has been removed." << std::endl;
					break;
				case IEE_EmoStateUpdated:
					IEE_EmoEngineEventGetEmoState(eEvent, eState);
					break;
				case IEE_MentalCommandEvent:
					handleMentalCommandEvent(userID, eEvent);
				case IEE_FacialExpressionEvent:

					break;
				default:
					break;
			}
		}
		else if (state != EDK_NO_EVENT) {
			std::cout << "Internal error in Emotiv Engine!" << std::endl;
			break;
		}

#ifdef _WIN32
    Sleep(5);
#endif
#if __linux__ || __APPLE__
        usleep(5000);
#endif
	}
    
	IEE_EngineDisconnect();
	IEE_EmoStateFree(eState);
	IEE_EmoEngineEventFree(eEvent);
    return -1;
}

void help()
{
    std::cout << "===================================================================" << std::endl;
    std::cout << "Please do follow steps:" << std::endl;
    std::cout << "A. Press 1 to Active training session." << std::endl;
    std::cout << "B. Press 2 to Start training NEUTRAL. When completed, press 3 ." << std::endl;
    std::cout << "C. Press 3 to Start training RIGHT. When completed, press 4 ." << std::endl;
    std::cout << "D. Press 4 to Start training LEFT." << std::endl;
    std::cout << "E. Press 5 to Set ActivationLevel." << std::endl;
    std::cout << "===================================================================" << std::endl;
    std::cout << "Otherwise, press key below for more: " << std::endl;
    std::cout << std::endl;
    std::cout << "Press 6 to Read ActivationLevel." << std::endl;
    std::cout << "Press 7 to Get OverallSkillRating." << std::endl;
    std::cout << "Press 8 to Save Profile Data to cloud." << std::endl;
    std::cout << "Press 9 to Load Profile Data from cloud." << std::endl;
    std::cout << std::endl;
    std::cout << "Press 0 for help." << std::endl;
    std::cout << "Press ` to quit." << std::endl;
    std::cout << "===================================================================" << std::endl;
}

#ifdef __linux__
int _kbhit(void)
{
    struct timeval tv;
    fd_set read_fd;

    tv.tv_sec=0;
    tv.tv_usec=0;

    FD_ZERO(&read_fd);
    FD_SET(0,&read_fd);

    if(select(1, &read_fd,NULL, NULL, &tv) == -1)
    return 0;

    if(FD_ISSET(0,&read_fd))
        return 1;

    return 0;
}

int _getch( void )
{
   struct termios oldattr, newattr;
   int ch;

   tcgetattr( STDIN_FILENO, &oldattr );
   newattr = oldattr;
   newattr.c_lflag &= ~( ICANON | ECHO );
   tcsetattr( STDIN_FILENO, TCSANOW, &newattr );
   ch = getchar();
   tcsetattr( STDIN_FILENO, TCSANOW, &oldattr );

   return ch;
}
#endif
#ifdef __APPLE__
int _kbhit (void)
{
    struct timeval tv;
    fd_set rdfs;

    tv.tv_sec = 0;
    tv.tv_usec = 0;

    FD_ZERO(&rdfs);
    FD_SET (STDIN_FILENO, &rdfs);

    select(STDIN_FILENO+1, &rdfs, NULL, NULL, &tv);
    return FD_ISSET(STDIN_FILENO, &rdfs);
}

int _getch( void )
{
    int r;
    unsigned char c;
    if((r = read(0, &c, sizeof(c))) < 0 )
    {
        return r;
    }
    else
    {
        return c;
    }
}
#endif

File: emotiv-sdk/examples_basic/Java/src/com/emotiv/examples/MentalCommandWithCloudProfile/MentalCommandWithCloudProfile.java
package com.emotiv.examples.MentalCommandWithCloudProfile;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import com.emotiv.Iedk.*;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.*;


public class MentalCommandWithCloudProfile implements Runnable {
	public static final Pointer eEvent = Edk.INSTANCE.IEE_EmoEngineEventCreate();
	public static final Pointer eState = Edk.INSTANCE.IEE_EmoStateCreate();
	public static BufferedReader in;

	public static int state = 0;
	public static IntByReference engineUserID = null;
	public static IntByReference userCloudID = null;
	public static IntByReference profileID = null;
	public static String profileName = null;
	public static int pending = 0;
	public static String option = null;
	
	@Override
	public void run() {
		while (true) {
			try {
				option = in.readLine();
				pending = 1;
				if (option.equals("q"))
					break;
			} catch (IOException e) {
				pending = 0;
			}
		}
	}

	public static void main(String[] args) throws InterruptedException {

		String userName = "Your EmotivID";
		String password = "Your password";
		profileName = "Your profile name";

		
		engineUserID = new IntByReference(0);
		userCloudID  = new IntByReference(0);
		profileID 	 = new IntByReference(-1);

		if (Edk.INSTANCE.IEE_EngineConnect("Emotiv Systems-5") != EdkErrorCode.EDK_OK.ToInt()) {
			System.out.println("Emotiv Engine start up failed.");
			return;
		}

		if (EmotivCloudClient.INSTANCE.EC_Connect() != EdkErrorCode.EDK_OK.ToInt()) {
			System.out.println("Cannot connect to Emotiv Cloud");
			return;
		}

		if (EmotivCloudClient.INSTANCE.EC_Login(userName, password) != EdkErrorCode.EDK_OK.ToInt()) {
			System.out.println("Your login attempt has failed. The username or password may be incorrect");
			return;
		}

		System.out.println("Logged in as " + userName);

		if (EmotivCloudClient.INSTANCE.EC_GetUserDetail(userCloudID) != EdkErrorCode.EDK_OK.ToInt()) {
			return;
		}

		help();
		
		in = new BufferedReader(new InputStreamReader(System.in));
		Thread t1 = new Thread(new MentalCommandWithCloudProfile());
		t1.start();
		
		startLiveClassificationProcess();
		
		System.out.println("Quitting...");

		Edk.INSTANCE.IEE_EngineDisconnect();
		Edk.INSTANCE.IEE_EmoStateFree(eState);
		Edk.INSTANCE.IEE_EmoEngineEventFree(eEvent);
		
		System.out.println("Wating for the Thread");

	}

	

	public static void startLiveClassificationProcess() {

		while (true) {
			if (pending == 1) {
				if(!handleUserInput(engineUserID, userCloudID, profileName, option))
					break;
				pending = 0;
			}

			state = Edk.INSTANCE.IEE_EngineGetNextEvent(eEvent);

			if (state == EdkErrorCode.EDK_OK.ToInt()) {
				int eventType = Edk.INSTANCE.IEE_EmoEngineEventGetType(eEvent);
				Edk.INSTANCE.IEE_EmoEngineEventGetUserId(eEvent, engineUserID);

				if (eventType == Edk.IEE_Event_t.IEE_UserAdded.ToInt())
					System.out.println("New user " + engineUserID.getValue() + " added");

				else if (eventType == Edk.IEE_Event_t.IEE_UserRemoved.ToInt())
					System.out.println("User " + engineUserID.getValue() + " has been removed.");

				else if (eventType == Edk.IEE_Event_t.IEE_EmoStateUpdated.ToInt()) {
					Edk.INSTANCE.IEE_EmoEngineEventGetEmoState(eEvent, eState); }

				else if (eventType == Edk.IEE_Event_t.IEE_MentalCommandEvent.ToInt()) {
					handleMentalCommandEvent(engineUserID, eEvent);
					System.out.println("MentalCommandEvent arrived");
				}
					
				else if (eventType == Edk.IEE_Event_t.IEE_FacialExpressionEvent.ToInt()) {
					System.out.println("FacialExpressionEvent arrived");
				}

				else {
				}

			} else if (state != EdkErrorCode.EDK_NO_EVENT.ToInt())
				System.out.println("Internal error in Emotiv Engine!");
		}

	}

	private static void SavingLoadingFunction(int userCloudID, int engineUserID, boolean save, String profileName) {
		int getNumberProfile = EmotivCloudClient.INSTANCE.EC_GetAllProfileName(userCloudID);

		int result = EmotivCloudClient.INSTANCE.EC_GetProfileId(userCloudID, profileName, profileID);

		if (save) {

			if (profileID.getValue() >= 0) {
				System.out.println("Profile with " + profileName + " exists.");
				if (EmotivCloudClient.INSTANCE.EC_UpdateUserProfile(userCloudID, engineUserID,
						profileID.getValue()) == EdkErrorCode.EDK_OK.ToInt()) {
					System.out.println("Updating finished");
				} else {
					System.out.println("updating failed");
				}
			} else if (EmotivCloudClient.INSTANCE.EC_SaveUserProfile(userCloudID, engineUserID, profileName,
					EmotivCloudClient.profileType_t.TRAINING.toInt()) == EdkErrorCode.EDK_OK.ToInt())
				System.out.println("Saving finished");
			else
				System.out.println("Saving failed");
			return;
		} else {
			if (getNumberProfile > 0) {
				if (EmotivCloudClient.INSTANCE.EC_LoadUserProfile(userCloudID, engineUserID, profileID.getValue(),
						-1) == EdkErrorCode.EDK_OK.ToInt())
					System.out.println("Loading finished");
				else
					System.out.println("Loading failed");
			}
			return;
		}

	}

	public static void help() {
		System.out.println("=============================================================");
		System.out.println("Please do follow steps:");
		System.out.println("A. Press 1 to Active training session.");
		System.out.println("B. Press 2 to Start training NEUTRAL. When completed, press 3.");
		System.out.println("C. Press 3 to Start training RIGHT. When completed, press 4");
		System.out.println("D. Press 4 to Start training LEFT.");
		System.out.println("E. Press 5 to Set ActivationLevel.");
		System.out.println("=============================================================");
		System.out.println("Otherwise, press key below for more:");
		System.out.println();
		System.out.println("Press 6 to Read ActivationLevel.");
		System.out.println("Press 7 to Get OveralSkillRating.");
		System.out.println("Press 8 to Save Profile Data to the cloud.");
		System.out.println("Press 9 to Load Profile Data from cloud.");
		System.out.println();
		System.out.println("Press 0 for help.");
		System.out.println("Press r to start live classification.");
		System.out.println("Press q to quit.");
		System.out.println("=============================================================");

	}

	private static void handleMentalCommandEvent(IntByReference engineUserID, Pointer cognitiveEvent) {
		int eventType = Edk.INSTANCE.IEE_MentalCommandEventGetType(cognitiveEvent);

		switch (eventType) {

		case 1: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " STARTED!");
			break;
		}

		case 2: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " SUCCEEDED!");
			System.out.println("Accept training!!!");
			Edk.INSTANCE.IEE_MentalCommandSetTrainingControl(engineUserID.getValue(),
					Edk.IEE_MentalCommandTrainingControl_t.MC_ACCEPT.getType());
			break;
		}

		case 3: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " FAILED!");
			break;
		}
		
		case 4: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " COMPLETED!");
			break;
		}

		case 5: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " ERASED!");
			break;
		}

		case 6: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " REJECTED!");
			break;
		}
		
		case 7: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " RESET!");
			break;
		}
		case 8: {
			System.out.println("MentalCommand training for user " + engineUserID.getValue() + " Auto Sampling Neutral Completed!");
			break;
		}

		case 0: {
			break;
		}

		default: {
			break;
		}
		}
	}
	
	public static boolean handleUserInput(IntByReference userId, IntByReference userCloudId, String profileName,
			String line) {
		switch (line) {
		case "1": {
			long action1 = (long) EmoState.IEE_MentalCommandAction_t.MC_RIGHT.ToInt();
			long action2 = (long) EmoState.IEE_MentalCommandAction_t.MC_LEFT.ToInt();
			long listAction = action1 | action2;
			int errorCode = EdkErrorCode.EDK_OK.ToInt();
			errorCode = Edk.INSTANCE.IEE_MentalCommandSetActiveActions(userId.getValue(), listAction);
			if (errorCode == EdkErrorCode.EDK_OK.ToInt())
				System.out.println(
						"Setting MentalCommand active actions (MC_LEFT | MC_RIGHT) for user " + userId.getValue());
			else
				System.out.println("Setting MentalCommand active actions error: " + errorCode);
			
			Edk.INSTANCE.IEE_EngineClearEventQueue(Edk.IEE_Event_t.IEE_EmoStateUpdated.ToInt());

			break;

		}
		case "2": {
			System.out.println("Starting training NEUTRAL!");
			Edk.INSTANCE.IEE_MentalCommandSetTrainingAction(userId.getValue(), EmoState.IEE_MentalCommandAction_t.MC_NEUTRAL.ToInt());
			Edk.INSTANCE.IEE_MentalCommandSetTrainingControl(userId.getValue(), Edk.IEE_MentalCommandTrainingControl_t.MC_START.getType());
			break;
		}
		case "3": {
			System.out.println("Starting training RIGHT!");
			Edk.INSTANCE.IEE_MentalCommandSetTrainingAction(userId.getValue(),
					EmoState.IEE_MentalCommandAction_t.MC_RIGHT.ToInt());
			Edk.INSTANCE.IEE_MentalCommandSetTrainingControl(userId.getValue(),
					Edk.IEE_MentalCommandTrainingControl_t.MC_START.getType());
			break;
		}
		case "4": {
			System.out.println("Starting training LEFT!");
			Edk.INSTANCE.IEE_MentalCommandSetTrainingAction(userId.getValue(),
					EmoState.IEE_MentalCommandAction_t.MC_LEFT.ToInt());
			Edk.INSTANCE.IEE_MentalCommandSetTrainingControl(userId.getValue(),
					Edk.IEE_MentalCommandTrainingControl_t.MC_START.getType());
			break;
		}
		case "5": {
			IntByReference level = new IntByReference(0);
			if (Edk.INSTANCE.IEE_MentalCommandSetActivationLevel(userId.getValue(), 2) == EdkErrorCode.EDK_OK.ToInt())
				Edk.INSTANCE.IEE_MentalCommandGetActivationLevel(userId.getValue(), level);
			else
				System.out.println("Set MentalComand Activation level failed " + level.getValue());
			break;
		}
		case "6": {
			IntByReference level = new IntByReference(0);
			Edk.INSTANCE.IEE_MentalCommandGetActivationLevel(userId.getValue(), level);
			System.out.println("Current MentalCommand Activation level: " + level.getValue());
		}
		case "7": {
			FloatByReference skill = new FloatByReference(0);
			Edk.INSTANCE.IEE_MentalCommandGetOverallSkillRating(userId.getValue(), skill);
			System.out.println("Current overall skill rating: " + skill.getValue());
			break;
		}
		case "8": {
			System.out.println("Saving profile...");
			SavingLoadingFunction(userCloudId.getValue(), userId.getValue(), true, profileName);
			break;
		}
		case "9": {
			System.out.println("Loading profile...");
			SavingLoadingFunction(userCloudId.getValue(), userId.getValue(), false, profileName);
			break;
		}
		case "0": {
			help();
			break;
		}
		case "q": {
			return false;
		}
		case "r": {
			startLiveClassificationProcess();
		}
		default: {
			break;
		}
		}
		return true;
	}
	
	public static boolean checkTrained(NativeLongByReference action, int userId) {
		int result = Edk.INSTANCE.IEE_MentalCommandGetTrainedSignatureActions(userId, action);
		if (result == EdkErrorCode.EDK_OK.ToInt()) {
			return true;
		}
		return false;
	}
}
